pkg/env-tests/scheduler_test.go

```diff
@@ -338,4 +369,132 @@ var _ = Describe("Scheduler", Ordered, func() {
+		})
+
+		AfterAll(func(ctx context.Context) {
+			if topologyTree != nil {
+				Expect(ctrlClient.Delete(ctx, topologyTree)).To(Succeed(), "Failed to delete topology tree")
+			}
+
+			for _, node := range topologyNodes {
+				Expect(ctrlClient.Delete(ctx, node)).To(Succeed(), "Failed to delete topology test node")
+			}
+
+			err := DeleteAllInNamespace(ctx, ctrlClient, testNamespace.Name,
+				&corev1.Pod{},
+				&schedulingv2alpha2.PodGroup{},
+				&resourcev1beta1.ResourceClaim{},
+				&kaiv1alpha2.BindRequest{},
+			)
+			Expect(err).NotTo(HaveOccurred(), "Failed to delete test resources")
+
+			err = WaitForNoObjectsInNamespace(ctx, ctrlClient, testNamespace.Name, defaultTimeout, interval,
+				&corev1.PodList{},
+				&schedulingv2alpha2.PodGroupList{},
+				&resourcev1beta1.ResourceClaimList{},
+				&kaiv1alpha2.BindRequestList{},
+			)
+			Expect(err).NotTo(HaveOccurred(), "Failed to wait for test resources to be deleted")
+		})
+
+		It("Schedule pods with rack topology constraints", func(ctx context.Context) {
+			singlePodResourceRequirements := corev1.ResourceRequirements{
+				Limits: corev1.ResourceList{
+					constants.GpuResource: resource.MustParse(fmt.Sprintf("%d", gpusPerNode)),
+				},
+				Requests: corev1.ResourceList{
+					constants.GpuResource: resource.MustParse(fmt.Sprintf("%d", gpusPerNode)),
+				},
+			}
+
+			numWorkloadPods := 4
+			workloadPods := []*corev1.Pod{}
+			for i := 0; i < numWorkloadPods; i++ {
+				testPod := CreatePodObject(testNamespace.Name, fmt.Sprintf("test-pod-%d", i), singlePodResourceRequirements)
+				Expect(ctrlClient.Create(ctx, testPod)).To(Succeed(), "Failed to create test pod %s", testPod.Name)
+				workloadPods = append(workloadPods, testPod)
+			}
+
+			podGroupConfig := podGroupConfig{
+				queueName:    testQueue.Name,
+				podgroupName: "test-podgroup",
+				minMember:    1,
+				topologyConstraint: &schedulingv2alpha2.TopologyConstraint{
+					Topology:              topologyName,
+					RequiredTopologyLevel: rackLabelKey,
+				},
+			}
+			err := GroupPods(ctx, ctrlClient, podGroupConfig, workloadPods)
+			Expect(err).NotTo(HaveOccurred(), "Failed to group pods")
+
+			for _, pod := range workloadPods {
+				err = WaitForPodScheduled(ctx, ctrlClient, pod.Name, testNamespace.Name, defaultTimeout, interval)
+				Expect(err).NotTo(HaveOccurred(), "Failed to wait for test pod to be scheduled")
+			}
+
+			pods := &corev1.PodList{}
+			Expect(ctrlClient.List(ctx, pods, client.InNamespace(testNamespace.Name))).
+				To(Succeed(), "Failed to list pods")
+
+			Expect(len(pods.Items)).To(Equal(numWorkloadPods), "Expected %d pods to be created in the test", numWorkloadPods)
+
+			scheduledRacks := map[string]int{}
+			for _, pod := range pods.Items {
+				scheduledRacks[pod.Labels[rackLabelKey]]++
+			}
+
+			Expect(scheduledRacks).To(HaveLen(1), "Expected all pods to be scheduled on the same rack. Got %v", scheduledRacks)
```

Comment:
But is there any reason they won't? Even without topology?
Feels like the tests should interfere, like somehow mess with the scoring of nodes, so the same scenario will fail without the topology.

Reply:
First, I think the tests should always be deterministic if possible - why not swap specific nodes?
Second, re-running the test just seems like a waste
What if you put 1 gpu-job on a node outside the rack so it will "pull" the pods due to bin packing?

Reply:
I implemented @itsomri suggestion